Fixes #27060 - add launch vm wait time setting

lzap opened the pull request:
Some backend engines (e.g. Infoblox DHCP) takes 1-2 minutes to apply the changes which leads to boot errors of VMs orchestrated by Foreman. New setting "Launch VM Wait Time" should solve this easily.

lzap committed:
Fixes #27060 - add launch vm wait time setting

timogoebel commented:
... doesn't this lock the database for this time as this is running in a transaction?

ohadlevy commented:
I wonder if it makes more sense to make the proxy api call blocking vs introducing a random delay?

ohadlevy commented:
or even a way to figure out from the proxy (a task?) when its done vs just randomly setting a delay? (e.g who guarantee next time it would enough?)

lzap commented:
> doesn't this lock the database for this time as this is running in a transaction?

Good point, however I don't believe we use explicit transactions here. We rarely use them.

> I wonder if it makes more sense

This is a workaround enabler, I am not aiming to refactor our orchestration or anything like that. We can't even tell when Infoblox finishes processing the request, the only way would be probably to perform DHCP fake/spoof requests until the expected data appear. Not sure if that's worth the effort.

I added check for transaction, if this would ever block transaction (e.g. in the future we put this into one) it would not perform this. Also moved the delay to the proper method - I accidentally added it to the method which creates a VM, however this was supposed to be before VM is powered up.

ekohl commented:
I agree with @ohadlevy that it should be solved in the right place. Perhaps the infoblox smart proxy plugin should actually do the sleep since that's the right place to detect it? It would also keep VMs and bare metal at the same level of support.

lzap commented:
Disagree, this is IMHO better place, let me explain. If we do it here, then Foreman UI will work just fine without any browser timeout since we actually do AJAX/JavaScript progress bar. However if we do it on smart proxy side, then Foreman will actually timeout as it defaults to 60 seconds. This is configurable in Administer Settings, however why not aim better user experience from the day one? In both cases, hammer CLI timeout must be adjusted in order to allow the transaction to finish.

This is more generic solution that will also help if users run into issues with other things in the transaction. So far we have heard about Infoblox but there are other complicated deployments - realms, other plugins. Are you gonna ask to put timeout also there? That does not sound right.

lzap commented:
What if I refactor this to be an extra orchestration step, so it's clearly visible there is an extra "wait step". I really think visibility is key here, hiding this in a smart proxy DNS module is sneaky. @ares

ekohl commented:
In university I was taught that sleep is never a proper solution. It's still a blind solution and it might break. What if the product changes the refresh interval? :-1: from me.

ares commented:
> I wonder if it makes more sense to make the proxy api call blocking vs introducing a random delay?

that would block all other proxy operations right? smart proxy is single thread app

> This is more generic solution that will also help if users run into issues with other things in the transaction. 

@lzap I don't like this workaround though. I understand that you try to fix it potentially for all other operations that may take too long. But I'd also prefer if the smart proxy plugin would hardcode the expected timeout or had its own setting. From user perspective, I wouldn't like to set generic "sleep" setting value because one of my subsystem may be slow.

I'd prefer if the smart proxy plugin would handle this async and we added a step checking that the DHCP writing has finished before we proceed. I suppose that would mean using smart_proxy_dynflow in smart_proxy_infoblox_dhcp. And I can understand it's a lot of work.

> We can't even tell when Infoblox finishes processing the request, the only way would be probably to perform DHCP fake/spoof requests until the expected data appear. Not sure if that's worth the effort.

If that is the case and there's no synchronous endpoint provided by Infoblox or no way to check, if the change is effective already, I'd still prefer to have this safe bumper as part of smart_proxy_infoblox_dhcp, so calling the write method, would simply block for 60 seconds. Which I don't like either, since it will block the proxy entirely. But that's probably the price of a workaround.

Did you speak with Infoblox devs how they would solve it?

timogoebel commented:
How about adding another orchestration step that checks that e.g. the DNS entry is actually present? Foreman can re-try a couple of times and eventually give up.

I don't like the current approach either and share @ares and @ekohl's concerns.

lzap commented:
> How about adding another orchestration step that checks that e.g. the DNS entry is actually present?

This is actually issue with DHCP, the entry is not appearing up in the Infoblox cluster for one or two minutes.

I've had enough, going for the dirty workaround in the Infoblox module.

ares commented:
@lzap I can see accepting something in the middle, if smart proxy dhcp module would expose timeout configuration (thanks to new capabilities feature) and the smart proxy infoblox dhcp would configure this to 2 minutes, Foreman orchestration could wait for such time in extra orchestration step. But it should be specific to the proxy configuration, not setting generic sleep for all proxies and all potential operations.

lzap commented:
I already implemented this in the Infoblox plugin: RM 27060 - post-restart grid delay
